File: android/src/main/java/org/ligi/passandroid/ui/PassViewActivityBase.java
package org.ligi.passandroid.ui;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewConfiguration;
import android.view.Window;
import android.view.WindowManager;
import java.io.IOException;
import java.lang.reflect.Field;
import javax.inject.Inject;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.ligi.passandroid.App;
import org.ligi.passandroid.R;
import org.ligi.passandroid.model.InputStreamWithSource;
import org.ligi.passandroid.model.Settings;
import org.ligi.passandroid.model.pass.Pass;
import org.ligi.passandroid.ui.UnzipPassController.InputStreamUnzipControllerSpec;

public class PassViewActivityBase extends PassAndroidActivity {

    public static final String EXTRA_KEY_UUID = "uuid";

    public Pass currentPass;
    private boolean fullBrightnessSet = false;

    @Inject
    Settings settings;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // a little hack because I strongly disagree with the style guide here
        // ;-)
        // not having the Actionbar overflow menu also with devices with hardware
        // key really helps discoverability
        // http://stackoverflow.com/questions/9286822/how-to-force-use-of-overflow-menu-on-devices-with-menu-button
        try {
            final ViewConfiguration config = ViewConfiguration.get(this);
            final Field menuKeyField = ViewConfiguration.class.getDeclaredField("sHasPermanentMenuKey");
            if (menuKeyField != null) {
                menuKeyField.setAccessible(true);
                menuKeyField.setBoolean(config, false);
            }
        } catch (Exception ex) {
            // Ignore - but at least we tried ;-)
        }

        App.component().inject(this);
    }

    @Override
    protected void onResume() {
        super.onResume();

        final String uuid = getIntent().getStringExtra(EXTRA_KEY_UUID);

        if (uuid != null) {
            final Pass passbookForId = passStore.getPassbookForId(uuid);
            passStore.setCurrentPass(passbookForId);
        }
        currentPass = passStore.getCurrentPass();

        if (currentPass == null) {
            tracker.trackException("pass not present in " + this, false);
            finish();
        }

        configureActionBar();

        if (settings.isAutomaticLightEnabled()) {
            setToFullBrightness();
        }
    }

    protected void configureActionBar() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    protected void refresh() {

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.activity_pass_view, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        final boolean res = super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.menu_light).setVisible(!fullBrightnessSet);
        menu.findItem(R.id.menu_print).setVisible(Build.VERSION.SDK_INT >= 19);
        return res;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (new PassMenuOptions(this, currentPass).process(item)) {
            return true;
        }

        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;

            case R.id.menu_light:
                setToFullBrightness();
                return true;

            case R.id.menu_update:
                new Thread(new UpdateAsync()).start();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public static boolean mightPassBeAbleToUpdate(Pass pass) {
        return pass != null && pass.getWebServiceURL() != null && pass.getPassIdent() != null && pass.getSerial() != null;
    }

    private class UpdateAsync implements Runnable {

        private ProgressDialog dlg;

        @Override
        public void run() {
            final Pass pass = currentPass;
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    dlg = new ProgressDialog(PassViewActivityBase.this);
                    dlg.setMessage(getString(R.string.downloading_new_pass_version));
                    dlg.show();
                }
            });

            final OkHttpClient client = new OkHttpClient();

            final String url = pass.getWebServiceURL() + "/v1/passes/" + pass.getPassIdent() + "/" + pass.getSerial();
            final Request.Builder requestBuilder = new Request.Builder().url(url);
            requestBuilder.addHeader("Authorization", "ApplePass " + pass.getAuthToken());

            final Request request = requestBuilder.build();

            final Response response;
            try {
                response = client.newCall(request).execute();
                final InputStreamWithSource inputStreamWithSource = new InputStreamWithSource(url, response.body().byteStream());
                final InputStreamUnzipControllerSpec spec = new InputStreamUnzipControllerSpec(inputStreamWithSource,
                                                                                               PassViewActivityBase.this,
                                                                                               passStore,
                                                                                               new MyUnzipSuccessCallback(dlg),
                                                                                               new MyUnzipFailCallback(dlg));
                spec.setOverwrite(true);
                UnzipPassController.INSTANCE.processInputStream(spec);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private class MyUnzipFailCallback implements UnzipPassController.FailCallback {
        private final Dialog dlg;

        private MyUnzipFailCallback(final Dialog dlg) {
            this.dlg = dlg;
        }

        @Override
        public void fail(@NonNull final String reason) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    if (!isFinishing()) {
                        dlg.dismiss();
                        new AlertDialog.Builder(PassViewActivityBase.this).setMessage("Could not update pass :( " + reason + ")")
                                                                          .setPositiveButton(android.R.string.ok, null)
                                                                          .show();
                    }
                }
            });

        }
    }

    private class MyUnzipSuccessCallback implements UnzipPassController.SuccessCallback {

        private final Dialog dlg;

        private MyUnzipSuccessCallback(final Dialog dlg) {
            this.dlg = dlg;
        }

        @Override
        public void call(@NonNull final String uuid) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    if (isFinishing()) {
                        return;
                    }
                    dlg.dismiss();
                    if (!currentPass.getId().equals(uuid)) {
                        passStore.deletePassWithId(currentPass.getId());
                    }
                    final Pass newPass = passStore.getPassbookForId(uuid);
                    passStore.setCurrentPass(newPass);
                    currentPass = passStore.getCurrentPass();
                    refresh();

                    Snackbar.make(getWindow().getDecorView(), R.string.pass_updated, Snackbar.LENGTH_LONG).show();
                }
            });

        }

    }

    private void setToFullBrightness() {
        final Window win = getWindow();
        final WindowManager.LayoutParams params = win.getAttributes();
        params.screenBrightness = WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_FULL;
        win.setAttributes(params);
        fullBrightnessSet = true;
        supportInvalidateOptionsMenu();
    }
}

File: android/src/main/java/org/ligi/passandroid/ui/PassViewActivityBase.kt
package org.ligi.passandroid.ui

import android.Manifest
import android.annotation.SuppressLint
import android.app.*
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.ViewConfiguration
import android.view.WindowManager
import android.widget.RemoteViews
import androidx.appcompat.app.AlertDialog
import androidx.core.app.NotificationCompat
import androidx.core.content.pm.ShortcutInfoCompat
import androidx.core.content.pm.ShortcutManagerCompat
import androidx.core.graphics.drawable.IconCompat
import androidx.core.graphics.scale
import com.google.android.material.snackbar.Snackbar
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.ligi.passandroid.R
import org.ligi.passandroid.model.InputStreamWithSource
import org.ligi.passandroid.model.PassBitmapDefinitions.BITMAP_ICON
import org.ligi.passandroid.model.State
import org.ligi.passandroid.model.pass.Pass
import org.ligi.passandroid.ui.UnzipPassController.InputStreamUnzipControllerSpec
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.RuntimePermissions
import permissions.dispatcher.ktx.constructPermissionsRequest
import java.io.IOException

@SuppressLint("Registered")
open class PassViewActivityBase : PassAndroidActivity() {

    lateinit var currentPass: Pass
    private var fullBrightnessSet = false
    private val NOTIFICATION_CHANNEL_ID = "passnotifications"
    val PASS_NOTIFICATION_ID = 5

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // a little hack because I strongly disagree with the style guide here
        // ;-)
        // not having the Actionbar overflow menu also with devices with hardware
        // key really helps discoverability
        // http://stackoverflow.com/questions/9286822/how-to-force-use-of-overflow-menu-on-devices-with-menu-button
        try {
            val config = ViewConfiguration.get(this)
            val menuKeyField = ViewConfiguration::class.java.getDeclaredField("sHasPermanentMenuKey")
            menuKeyField.isAccessible = true
            menuKeyField.setBoolean(config, false)
        } catch (ex: Exception) {
            // Ignore - but at least we tried ;-)
        }

        updateCurrentPass()
    }

    override fun onPause() {
        super.onPause()
        State.lastSelectedPassUUID = currentPass.id
    }

    override fun onResume() {
        super.onResume()

        configureActionBar()

        if (settings.isAutomaticLightEnabled()) {
            setToFullBrightness()
        }
    }

    private fun updateCurrentPass() {
        val uuid = intent.getStringExtra(EXTRA_KEY_UUID)

        if (uuid != null) {
            passStore.currentPass = passStore.getPassbookForId(uuid)
        }

        if (passStore.currentPass == null) {
            passStore.currentPass = passStore.getPassbookForId(State.lastSelectedPassUUID)
        }

        if (passStore.currentPass == null) {
            tracker.trackException("pass not present in $this", false)
            finish()
            return
        }

        currentPass = passStore.currentPass!!
    }

    protected fun configureActionBar() {
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    protected open fun refresh() {
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_pass_view, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val res = super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.menu_light).isVisible = !fullBrightnessSet
        menu.findItem(R.id.menu_print).isVisible = Build.VERSION.SDK_INT >= 19
        return res
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (PassMenuOptions(this, currentPass).process(item)) {
            return true
        }

        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }

            R.id.menu_light -> {
                setToFullBrightness()
                true
            }

            R.id.install_shortcut -> {
                createShortcut()
                true
            }

            R.id.menu_update -> {
                Thread(UpdateAsync()).start()
                true
            }
            R.id.menu_notification -> {
                createNotification()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    }

    private fun createShortcut() {
        constructPermissionsRequest(Manifest.permission.INSTALL_SHORTCUT) {
            val passBitmap = currentPass.getBitmap(passStore, BITMAP_ICON)
            val shortcutIcon = passBitmap?.scale(128, 128, filter = true) ?: BitmapFactory.decodeResource(resources, R.drawable.ic_launcher)
            val name: CharSequence = currentPass.description.let {
                if (it.isNullOrEmpty()) "pass" else it
            }

            val targetIntent = Intent(this, PassViewActivity::class.java)
                .setAction(Intent.ACTION_MAIN)
                .putExtra(EXTRA_KEY_UUID, currentPass.id)
            val shortcutInfo = ShortcutInfoCompat.Builder(this, "shortcut$name")
                .setIntent(targetIntent)
                .setShortLabel(name)
                .setIcon(IconCompat.createWithBitmap(shortcutIcon))
                .build()
            ShortcutManagerCompat.requestPinShortcut(this, shortcutInfo, null)
        }.launch()
    }

    private fun createNotification() {
        val expandedView = RemoteViews(packageName, R.layout.notification_expanded)
        val notifyManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (Build.VERSION.SDK_INT > 25) {
            val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, "PassAndroid Pass Notification", NotificationManager.IMPORTANCE_DEFAULT)
            channel.description = "Notifications to quickly show and open PassAndroid passes"
            notifyManager.createNotificationChannel(channel)
        }

        val passNotificationBuilder = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
        passNotificationBuilder.setSmallIcon(R.drawable.ic_theaters_notification)
                .setContentTitle(currentPass.description)
                .setContentText(getString(R.string.notification_text))

        val iconBitmap = currentPass.getBitmap(passStore, BITMAP_ICON)
        if (iconBitmap != null) {
            passNotificationBuilder.setLargeIcon(iconBitmap)
        }

        val barCodeBitmap = currentPass.barCode?.getBitmap(this.resources)!!.bitmap
        if (barCodeBitmap != null) {
            val height = (240 * this.resources.displayMetrics.density).toInt()
            val originalHeight = barCodeBitmap.height
            val originalWidth = barCodeBitmap.width
            val scale = height / originalHeight
            val width = (originalWidth * scale).toInt()
            val scaledBarCodeBitmap = Bitmap.createScaledBitmap(barCodeBitmap, height, width, false)
            expandedView.setImageViewBitmap(R.id.image_view_expanded, scaledBarCodeBitmap)
            passNotificationBuilder.setCustomBigContentView(expandedView)
                    .setStyle(NotificationCompat.DecoratedCustomViewStyle())
        }

        val intent = Intent(this, PassViewActivity::class.java)
        val pendingIntent: PendingIntent = PendingIntent.getActivity(this, 0, intent, 0)
        passNotificationBuilder.setContentIntent(pendingIntent)
        passNotificationBuilder.setAutoCancel(true)
        notifyManager.notify(PASS_NOTIFICATION_ID, passNotificationBuilder.build())
    }

    inner class UpdateAsync : Runnable {

        private lateinit var dlg: ProgressDialog

        override fun run() {
            val pass = currentPass
            runOnUiThread {
                dlg = ProgressDialog(this@PassViewActivityBase)
                dlg.setMessage(getString(R.string.downloading_new_pass_version))
                dlg.show()
            }

            val client = OkHttpClient()

            val url = pass.webServiceURL + "/v1/passes/" + pass.passIdent + "/" + pass.serial
            val requestBuilder = Request.Builder().url(url)
            requestBuilder.addHeader("Authorization", "ApplePass " + pass.authToken)

            val request = requestBuilder.build()

            val response: Response
            try {
                response = client.newCall(request).execute()
                val body = response.body()
                if (body != null) {
                    val inputStreamWithSource = InputStreamWithSource(url, body.byteStream())
                    val spec = InputStreamUnzipControllerSpec(inputStreamWithSource, this@PassViewActivityBase, passStore,
                            MyUnzipSuccessCallback(dlg),
                            MyUnzipFailCallback(dlg))
                    spec.overwrite = true
                    UnzipPassController.processInputStream(spec)
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    inner class MyUnzipFailCallback constructor(private val dlg: Dialog) : UnzipPassController.FailCallback {

        override fun fail(reason: String) {
            runOnUiThread {
                if (!isFinishing) {
                    dlg.dismiss()
                    AlertDialog.Builder(this@PassViewActivityBase).setMessage("Could not update pass :( $reason)")
                            .setPositiveButton(android.R.string.ok, null)
                            .show()
                }
            }

        }
    }

    inner class MyUnzipSuccessCallback constructor(private val dlg: Dialog) : UnzipPassController.SuccessCallback {

        override fun call(uuid: String) {
            runOnUiThread(Runnable {
                if (isFinishing) {
                    return@Runnable
                }
                dlg.dismiss()
                if (currentPass.id != uuid) {
                    passStore.deletePassWithId(currentPass.id)
                }
                val newPass = passStore.getPassbookForId(uuid)
                passStore.currentPass = newPass
                currentPass = passStore.currentPass!!
                refresh()

                Snackbar.make(window.decorView, R.string.pass_updated, Snackbar.LENGTH_LONG).show()
            })

        }

    }

    private fun setToFullBrightness() {
        val win = window
        val params = win.attributes
        params.screenBrightness = WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_FULL
        win.attributes = params
        fullBrightnessSet = true
        invalidateOptionsMenu()
    }

    companion object {

        const val EXTRA_KEY_UUID = "uuid"

        fun mightPassBeAbleToUpdate(pass: Pass?): Boolean {
            return pass?.webServiceURL != null && pass.passIdent != null && pass.serial != null
        }
    }
}